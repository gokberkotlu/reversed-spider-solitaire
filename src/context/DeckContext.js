import React, { useState, createContext, useEffect } from "react";
import { cardSeries } from '../utils/CardSeries';

export const DeckContext = createContext();

export const DeckProvider = props => {

    // States
    const [gameCards, setGameCards] = useState([[],[],[],[],[],[],[],[],[],[]]); // Represents cards on table
    const [remainingCards, setRemainingCards] = useState([]); // Cards outside the game
    const [numberOfCompleted, setNumberOfCompleted] = useState(Array(0)); // Shows number of completed series
    const [newCardPacks, setNewCardPacks] = useState(Array(5)); // Shows number of new card packs
    const [isGameStarted, setIsGameStarted] = useState(false); // Set state when any card action happened
    const [statesLog, setStatesLog] = useState([]); // Log States to Undo the Game

    // States: Score Datas
    const [score, setScore] = useState(500); // Game score which starts 500
    const [highScore, setHighScore] = useState(+localStorage.getItem('highScore')); // Recorded highscore to store in localStorage

    const dealTheCards = (cards, numberOfCards) => {
        
        // Save Changes When the New Card Pack Added
        logGameChanges();

        if(cards.length >= 10) {
            let gameSet = [...gameCards];
            for(let i = 0; i < numberOfCards; i++) {
                gameSet[i%10].push(cards[i]);
            }
            setGameCards(gameSet);
            setRemainingCards(cards.splice(numberOfCards));

            return gameSet;
        }
    }

    // arrangeReversedCards arranges the reversed card at the start
    const arrangeReversedCards = (gameSet) => {
        gameSet.forEach(cards => {
            for(let i = 0; i < cards.length - 1; i++) {
                cards[i].isCardReverse = true;
            }
        })
    }

    // suffleTheCards shuffles eight decks of cards
    const suffleTheCards = () => {
        let deskNumber = 8;
        let newArray = [];
        for(let i = 0; i < deskNumber; i++) {
            newArray = newArray.concat(
                [...cardSeries.map((card) => {
                    return {
                        "cardNum": card,
                        "cardDeck": (i + 1),
                        "isCardReverse": false
                    }
                })]);
        }
        var currentIndex = newArray.length,
            randomIndex;
        // While there remain elements to shuffle...
        while (0 !== currentIndex) {
            // Pick a remaining element...
            randomIndex = Math.floor(Math.random() * currentIndex);
            currentIndex--;
    
            // And swap it with the current element.
            [newArray[currentIndex], newArray[randomIndex]] = [
                newArray[randomIndex], newArray[currentIndex]
            ];
        }
        return newArray;
    }

    // logGameChanges takes snapshots before all card actions and stores them inside statesLog
    const logGameChanges = () => {
        let addToLog = {
            "gameCardsBefore": JSON.parse(JSON.stringify(gameCards)),
            "remainingCardsBefore": JSON.parse(JSON.stringify(remainingCards)),
            "numberOfCompletedBefore": JSON.parse(JSON.stringify(numberOfCompleted)),
            "newCardPacksBefore": JSON.parse(JSON.stringify(newCardPacks))
        }
        setStatesLog(prevState => [...prevState, addToLog]);
    }

    // undoTheGame sets game states with last logged snapshot
    const undoTheGame = () => {
        let statesLogValue = [...statesLog];
        let statesLogValueLength = statesLogValue.length
        if(statesLogValueLength > 1) {
            let beforeStates = statesLogValue[statesLogValueLength - 1]; // Go back in the game by having "before value of last screen"
            setGameCards(beforeStates.gameCardsBefore);
            setRemainingCards(beforeStates.remainingCardsBefore);
            setNumberOfCompleted(beforeStates.numberOfCompletedBefore);
            setNewCardPacks(beforeStates.newCardPacksBefore);

            // Delete current states from log when went back in the game
            setStatesLog(statesLogValue.slice(0, statesLogValueLength - 1));

            // Decrease 1 point from score for each undo click
            setScore(prevState => prevState - 1);

            // Decrease 100 points from score if completed desk is removed by clicking undo button
            if(beforeStates.numberOfCompletedBefore.length < numberOfCompleted.length) {
                let scoreDiff = numberOfCompleted.length - beforeStates.numberOfCompletedBefore.length;
                setScore(prevState => prevState - 100 * scoreDiff);
            }
        }
    }

    useEffect(() => {
        // Shuffle cards, arrange reversed cards and deal them on the game table at start
        var allCards = suffleTheCards();
        var gameSet = dealTheCards(allCards, 54);
        arrangeReversedCards(gameSet);
    }, [])

    return (
        <DeckContext.Provider
        value={{
                "gameCards": [gameCards, setGameCards],
                "remainingCards": [remainingCards, setRemainingCards],
                "numberOfCompleted": [numberOfCompleted, setNumberOfCompleted],
                "newCardPacks": [newCardPacks, setNewCardPacks],
                "isGameStarted": [isGameStarted, setIsGameStarted],
                "suffleTheCards": suffleTheCards,
                "dealTheCards": dealTheCards,
                "arrangeReversedCards": arrangeReversedCards,
                "statesLog": [statesLog, setStatesLog],
                "logGameChanges": logGameChanges,
                "undoTheGame": undoTheGame,
                "score": [score, setScore],
                "highScore": [highScore, setHighScore]
            }}>
            {props.children}
        </DeckContext.Provider>
    );
}