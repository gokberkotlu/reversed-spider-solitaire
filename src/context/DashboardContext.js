import React, { useState, createContext, useContext, useEffect } from "react";
import { DeckContext } from "./DeckContext";

export const DashboardContext = createContext();

export const DashboardProvider = props => {

    // States
    const [counter, setCounter] = useState("00:00:00"); // Represents timer
    const [isGameFinished, setIsGameFinished] = useState(false); // According to this state, popup is shown or not
    const [resetBit, setResetBit] = useState(false); // It triggers the useEffect to reset the game
    const [decreaseInEveryTenSeconds, setDecreaseInEveryTenSeconds] = useState(0); // Decrease 2 points from score for each 10 sec

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [, setGameCards] = deckContextValue.gameCards;
    const [, setRemainingCards] = deckContextValue.remainingCards;
    const [, setNumberOfCompleted] = deckContextValue.numberOfCompleted;
    const [, setNewCardPacks] = deckContextValue.newCardPacks;
    const [, setIsGameStarted] = deckContextValue.isGameStarted;
    const [, setStatesLog] = deckContextValue.statesLog;
    const suffleTheCards = deckContextValue.suffleTheCards;
    const dealTheCards = deckContextValue.dealTheCards;
    const arrangeReversedCards = deckContextValue.arrangeReversedCards;

    const [, setScore] = deckContextValue.score;

    const resetGame = () => {
        // Resetting
        setGameCards([[],[],[],[],[],[],[],[],[],[]]);
        setRemainingCards([]);
        setNewCardPacks(Array(5));
        setNumberOfCompleted(Array(0));
        setScore(500);
        setCounter("00:00:00");
        setIsGameFinished(false);
        setResetBit(true);
        setIsGameStarted(false);
        setStatesLog([]);
        setDecreaseInEveryTenSeconds(0);
    }

    useEffect(() => {// Deal cards again from scratch
        if(resetBit) {
            var allCards = suffleTheCards();
            var gameSet = dealTheCards(allCards, 54); // setGameCards and setRemainingCards included
            arrangeReversedCards(gameSet);
            setResetBit(false);
        }
    }, [resetBit]) // resetBit changes trigger the control of gameSet and remainingCards

    return (
        <DashboardContext.Provider
        value={{"counter": [counter, setCounter],
        "isGameFinished": [isGameFinished, setIsGameFinished],
        "resetGame": resetGame,
        "decreaseInEveryTenSeconds": [decreaseInEveryTenSeconds ,setDecreaseInEveryTenSeconds]}}>
            {props.children}
        </DashboardContext.Provider>
    );
}