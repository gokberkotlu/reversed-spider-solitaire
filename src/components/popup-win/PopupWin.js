import { useContext } from "react";
import { DeckContext } from "../../context/DeckContext";
import { DashboardContext } from "../../context/DashboardContext";
import "./popup-win.css";

const PopupWin = () => {

    // DashboardContext
    const dashboardContextValue = useContext(DashboardContext);

    const [isGameFinished, setIsGameFinished] = dashboardContextValue.isGameFinished;
    const resetGame = dashboardContextValue.resetGame;

    // DeckContext
    const deckContextValue = useContext(DeckContext);

    const [score,] = deckContextValue.score;

    return (
        (isGameFinished && <>
            <div className="popup-win">
                <div className="popup-close-btn" onClick={() => {
                    setIsGameFinished(false);
                }}>
                    <div className="bar"></div>
                </div>
                <div>You Win!</div>
                <button className="popup-reset-btn" onClick={() => {
                    resetGame();
                }}>New Game</button>
                <p className="popup-score">
                    <span className="score-bold">Score: </span>{score}
                </p>
            </div>
            <div className="blur-container"></div>
        </>)
    );
}
 
export default PopupWin;