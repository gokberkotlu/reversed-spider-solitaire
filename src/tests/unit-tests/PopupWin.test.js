import React from 'react';
import { screen, render } from '@testing-library/react';
import renderer from "react-test-renderer";
import PopupWin from "../../components/popup-win/PopupWin";
import { DeckProvider } from '../../context/DeckContext';
import { DashboardContext, DashboardProvider } from "../../context/DashboardContext";

describe('PopupWin component', () => {

    it('should render PopupWin component', () => {
        const tree = renderer.create(
            <DeckProvider>
                <DashboardProvider>
                    <PopupWin />
                </DashboardProvider>
            </DeckProvider>
        ).toJSON();
        expect(tree).toMatchSnapshot();
    })

    it('should have text content You Win! when game is finished', () => {
        render(
            <DeckProvider>
                <DashboardContext.Provider
                value={{
                    "isGameFinished": [true, null]
                }}>
                    <PopupWin />
                </DashboardContext.Provider>
            </DeckProvider>
        );
        expect(screen.getByText("You Win!")).toBeTruthy();
    })

    it('should have new game button inside popup when game is finished', () => {
        render(
            <DeckProvider>
                <DashboardContext.Provider
                value={{
                    "isGameFinished": [true, null]
                }}>
                    <PopupWin />
                </DashboardContext.Provider>
            </DeckProvider>
        );
        const newGameButton = screen.getByRole('button', { name: 'New Game' });
        expect(newGameButton).toBeTruthy();
    })
})