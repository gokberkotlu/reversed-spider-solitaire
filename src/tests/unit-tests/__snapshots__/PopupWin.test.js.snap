// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PopupWin component should render PopupWin component 1`] = `null`;
